Mathlib/AlgebraicTopology/RelativeCellComplex/AttachCells.lean

```diff
@@ -0,0 +1,164 @@
+```
+⨿ i, A (π i) -----> X₁
+  |                 |f
+  v                 v
+⨿ i, B (π i) -----> X₂
+```
+In other words, the morphism `f` is a pushout of coproducts of morphisms
+of the form `g a : A a ⟶ B a`, see `nonempty_attachCells_iff`.
+
+-/
+
+universe w t t' v u
+
+open CategoryTheory Limits
+
+namespace HomotopicalAlgebra
+
+variable {C : Type u} [Category.{v} C]
+  {α : Type t} {A B : α → C} (g : ∀ a, A a ⟶ B a)
+  {X₁ X₂ : C} (f : X₁ ⟶ X₂)
+
+/-- Given a family of morphisms `g a : A a ⟶ B a` and a morphism `f : X₁ ⟶ X₂`,
+this structure contains the data and properties which expresses that `X₂`
+is obtained from `X₁` by attaching cells of the form `g a`. -/
+structure AttachCells where
+  /-- the index type of the cells -/
+  ι : Type w
+  /-- for each `i : ι`, we shall attach a cell given by the morphism `g (π i)`. -/
+  π : ι → α
+  /-- a colimit cofan which gives the coproduct of the object `A (π i)` -/
+  cofan₁ : Cofan (fun i ↦ A (π i))
+  /-- a colimit cofan which gives the coproduct of the object `B (π i)` -/
+  cofan₂ : Cofan (fun i ↦ B (π i))
+  /-- `cofan₁` is colimit -/
+  isColimit₁ : IsColimit cofan₁
+  /-- `cofan₂` is colimit -/
+  isColimit₂ : IsColimit cofan₂
+  /-- the coproduct of the maps `g (π i) : A (π i) ⟶ B (π i)` for all `i : ι`. -/
+  m : cofan₁.pt ⟶ cofan₂.pt
+  hm (i : ι) : cofan₁.inj i ≫ m = g (π i) ≫ cofan₂.inj i := by aesop_cat
+  /-- the top morphism of the pushout square -/
+  g₁ : cofan₁.pt ⟶ X₁
+  /-- the bottom morphism of the pushout square -/
+  g₂ : cofan₂.pt ⟶ X₂
+  isPushout : IsPushout g₁ m f g₂
+
+namespace AttachCells
+
+open MorphismProperty
+
+attribute [reassoc (attr := simp)] hm
+
+variable {g f} (c : AttachCells.{w} g f)
+
+include c
+
+lemma pushouts_coproducts : (coproducts.{w} (ofHoms g)).pushouts f := by
+  refine ⟨_, _, _, _, _, ?_, c.isPushout⟩
+  have : c.m = c.isColimit₁.desc
+      (Cocone.mk _ (Discrete.natTrans (fun ⟨i⟩ ↦ by exact g (c.π i)) ≫ c.cofan₂.ι)) :=
+    c.isColimit₁.hom_ext (fun ⟨i⟩ ↦ by rw [IsColimit.fac]; exact c.hm i)
+  rw [this, coproducts_iff]
+  exact ⟨c.ι, ⟨_, _, _, _, c.isColimit₁, c.isColimit₂, _, fun i ↦ ⟨_⟩⟩⟩
+
+/-- The inclusion of a cell. -/
+def cell (i : c.ι) : B (c.π i) ⟶ X₂ := c.cofan₂.inj i ≫ c.g₂
+
+@[reassoc]
+lemma cell_def (i : c.ι) : c.cell i = c.cofan₂.inj i ≫ c.g₂ := rfl
+
+lemma hom_ext {Z : C} {φ φ' : X₂ ⟶ Z}
+    (h₀ : f ≫ φ = f ≫ φ') (h : ∀ i, c.cell i ≫ φ = c.cell i ≫ φ') :
+    φ = φ' := by
+  apply c.isPushout.hom_ext h₀
+  apply Cofan.IsColimit.hom_ext c.isColimit₂
+  simpa [cell_def] using h
+
+section
+
+variable {α' : Type t'} {A' B' : α' → C} (g' : ∀ i', A' i' ⟶ B' i')
+  (a : α → α') (ha : ∀ (i : α), Arrow.mk (g i) ≅ Arrow.mk (g' (a i)))
+
+/-- If a family of maps `g` is contained in another family `g'` (up to isomorphisms),
+if `f : X₁ ⟶ X₂` is a morphism, and `X₂` is obtained from `X₁` by attaching cells
+of the form `g`, then it is also obtained by attached cells of the form `g'`. -/
+def chg : AttachCells g' f where
```

Comment:
Is there a better name for this? Maybe `reindexCellTypes`?

Reply:
Thanks for the suggestion!